best_practices: encourage use of pylint, six, and python-modernize

davvid opened the pull request:
Add a section describing the Conservative Python 3 Porting Guide,
the python-futures Python 3 Compatibility Cheat Sheet, and the
python-modernize tool.

Recommend use of pylint --py3k.
Recommend use of six and python-modernize.
Add guidance for futurize and the python-future modules.

Signed-off-by: David Aguilar <davvid@gmail.com>

davvid committed:
best_practices: encourage use of pylint, six, and python-modernize

Add a section describing the Conservative Python 3 Porting Guide,
and the python-modernize tool.

Recommend use of pylint --py3k.
Recommend use of six and python-modernize.
Add guidance for futurize and the python-future modules.

Signed-off-by: David Aguilar <davvid@gmail.com>

AWhetter commented on deliverables/best_practices.md:
```diff
@@ -125,3 +125,106 @@ Always choose the version of Python that is correct for your usage.
+## Read the futurize Python 3 Compatibility Cheat Sheet
+
+The futurize project contains a
+[Cheat Sheet](https://python-future.org/compatible_idioms.html)
```
This guide and the above porting guide cover the same thing for the most part. Perhaps instead of recommending reading both we suggest one or the other depending on the situation?
The porting guide is much better for reading and learning from I think, whereas the future guide is a simple list of everything and is more useful for reference (eg linking to and doing a quick find).

davvid replied:
That's a good point.  Maybe adding the futurize docs to the mix here is actually making things too complicated since we would then have to explain the situation where the futurize docs may or may not be appropriate.

It seems like the simplest thing would be to recommend the CPPG and maybe link to the futurize cheat sheet in a separate "Additional Links" section at the end.  That way we acknowledge it, but our main recommendation becomes the CPPG.

AWhetter replied:
That sounds like a good approach to me. I like it!

AWhetter commented on deliverables/best_practices.md:
```diff
@@ -125,3 +125,106 @@ Always choose the version of Python that is correct for your usage.
+
+Pylint contains a "py3k" mode that enables Python 2 + 3 compatibility checks
+and is extremely helpful when transitioning to Python 2 + 3 compatibility.
+Use of `pylint --py3k` is strongly encouraged.
```
It's also useful as a way of ensuring that Python 3 compatibility remains in a repository, especially when there aren't many tests. Is this worth mentioning?

AWhetter commented on deliverables/best_practices.md:
```diff
@@ -125,3 +125,106 @@ Always choose the version of Python that is correct for your usage.
+## Consider use of futurize stage 1
+
+While we do not recommend the use of `futurize --stage2`, particularly for
+library code, `futurize --stage1` fixes are safe and can be a helpful tool.
```
How does this interact with modernize? I've never used the two together.

davvid replied:
Ditto for making this simpler -- perhaps we should only mention modernize and drop the complications around trying to use futurize stage1.  I'm not sure how they interact, and I've only used futurize on a few limited toy examples, so it seems better to drop this section altogether.

davvid commented:
Thanks for the review!  I added a note about pylint for projects that don't have tests and have adjusted the futurize links and recommendations as discussed.  Please see the latest pushed commit.

davvid committed:
deliverables: rename best_practices.md to transition_guide.md

Reposition the "Best Practices" document as the
"Python 3 Transition Guide" deliverable.

Signed-off-by: David Aguilar <davvid@gmail.com>

JeanChristopheMorinPerso commented:
Regarding the usage of `six`, should we mention that DCCs might ship with an older version of `six` and that the DCC version will always be picked up? On the top of my mind there is Nuke and Houdini that both ship a quite old `six` version.

This has bitten RodeoFX quite a lot. RodeoFX contacted both vendors and their response was that they would fix it in their next major version. Though I don't know where they are at right now.

AWhetter commented:
That's definitely been a problem for us as well. I think it's worth mentioning. Given the age of this pull request, I'm going to go ahead and merge this but I've made not of it as a point of discussion for the next meeting. Feel free to make a PR for this change ahead of time if you want.

JeanChristopheMorinPerso commented:
Perfect @AWhetter . I'll create a PR later today.
